In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns



In [2]:
# load the dataset
df = pd.read_csv("Alphabets_data.csv")
df.head()

,letter,xbox,ybox,width,height,onpix,xbar,ybar,x2bar,y2bar,xybar,x2ybar,xy2bar,xedge,xedgey,yedge,yedgex
0,T,2,8,3,5,1,8,13,0,6,6,10,8,0,8,0,8
1,I,5,12,3,7,2,10,5,5,4,13,3,9,2,8,4,10
2,D,4,11,6,8,6,10,6,2,6,10,3,7,3,7,3,9
3,N,7,11,6,6,3,5,9,4,6,4,4,10,6,10,2,8
4,G,2,1,3,1,1,8,6,6,6,6,5,9,1,7,5,10


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 17 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   letter  20000 non-null  object
 1   xbox    20000 non-null  int64 
 2   ybox    20000 non-null  int64 
 3   width   20000 non-null  int64 
 4   height  20000 non-null  int64 
 5   onpix   20000 non-null  int64 
 6   xbar    20000 non-null  int64 
 7   ybar    20000 non-null  int64 
 8   x2bar   20000 non-null  int64 
 9   y2bar   20000 non-null  int64 
 10  xybar   20000 non-null  int64 
 11  x2ybar  20000 non-null  int64 
 12  xy2bar  20000 non-null  int64 
 13  xedge   20000 non-null  int64 
 14  xedgey  20000 non-null  int64 
 15  yedge   20000 non-null  int64 
 16  yedgex  20000 non-null  int64 
dtypes: int64(16), object(1)
memory usage: 2.6+ MB


In [4]:
df.describe()

,xbox,ybox,width,height,onpix,xbar,ybar,x2bar,y2bar,xybar,x2ybar,xy2bar,xedge,xedgey,yedge,yedgex
count,20000.000000,20000.000000,20000.000000,20000.00000,20000.000000,20000.000000,20000.000000,20000.000000,20000.000000,20000.000000,20000.00000,20000.000000,20000.000000,20000.000000,20000.000000,20000.00000
mean,4.023550,7.035500,5.121850,5.37245,3.505850,6.897600,7.500450,4.628600,5.178650,8.282050,6.45400,7.929000,3.046100,8.338850,3.691750,7.80120
std,1.913212,3.304555,2.014573,2.26139,2.190458,2.026035,2.325354,2.699968,2.380823,2.488475,2.63107,2.080619,2.332541,1.546722,2.567073,1.61747
min,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000,0.00000
25%,3.000000,5.000000,4.000000,4.00000,2.000000,6.000000,6.000000,3.000000,4.000000,7.000000,5.00000,7.000000,1.000000,8.000000,2.000000,7.00000
50%,4.000000,7.000000,5.000000,6.00000,3.000000,7.000000,7.000000,4.000000,5.000000,8.000000,6.00000,8.000000,3.000000,8.000000,3.000000,8.00000
75%,5.000000,9.000000,6.000000,7.00000,5.000000,8.000000,9.000000,6.000000,7.000000,10.000000,8.00000,9.000000,4.000000,9.000000,5.000000,9.00000
max,15.000000,15.000000,15.000000,15.00000,15.000000,15.000000,15.000000,15.000000,15.000000,15.000000,15.00000,15.000000,15.000000,15.000000,15.000000,15.00000


In [5]:
df.shape

(20000, 17)

In [6]:
# checking for missing values:
df.isnull().sum()

letter    0
xbox      0
ybox      0
width     0
height    0
onpix     0
xbar      0
ybar      0
x2bar     0
y2bar     0
xybar     0
x2ybar    0
xy2bar    0
xedge     0
xedgey    0
yedge     0
yedgex    0
dtype: int64

In [7]:
# checking for duplicates:
df.duplicated().sum()

np.int64(1332)

In [8]:
df.drop_duplicates(inplace= True)

In [9]:
df.duplicated().sum()

np.int64(0)

In [10]:
# split the data into target and predictors:

y = df['letter']
x = df.drop('letter', axis=1)


In [11]:
# after spliting convert the categorical variable to numerical using label encoding

from sklearn.preprocessing import LabelEncoder, StandardScaler

le = LabelEncoder()
y = le.fit_transform(y)

## 2. Model Implementation

In [12]:
# split the dataset to train and test:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(
    x, y, test_size = 0.2,random_state = 42, stratify = y
)


In [13]:
# normalize :
scaler = StandardScaler()
x_train = scaler.fit_transform(x_train)
x_test = scaler.transform(x_test)

In [14]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.optimizers import Adam

In [15]:
# Define model 
model = Sequential([
    Dense(64, activation='relu', input_shape=(x_train.shape[1],)),  
    Dense(len(np.unique(y)), activation='softmax')                  
])

/opt/anaconda3/envs/tfenv/lib/python3.10/site-packages/keras/src/layers/core/dense.py:95: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [17]:
# compile model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [18]:
history = model.fit(x_train, y_train,
                    epochs=30,
                    batch_size=32,
                    validation_split=0.2,
                    verbose=1)

Epoch 1/30
374/374 ━━━━━━━━━━━━━━━━━━━━ 1s 714us/step - accuracy: 0.4076 - loss: 2.2402 - val_accuracy: 0.6307 - val_loss: 1.5066
Epoch 2/30
374/374 ━━━━━━━━━━━━━━━━━━━━ 0s 505us/step - accuracy: 0.6828 - loss: 1.2191 - val_accuracy: 0.7218 - val_loss: 1.0655
Epoch 3/30
374/374 ━━━━━━━━━━━━━━━━━━━━ 0s 507us/step - accuracy: 0.7489 - loss: 0.9292 - val_accuracy: 0.7529 - val_loss: 0.8980
Epoch 4/30
374/374 ━━━━━━━━━━━━━━━━━━━━ 0s 505us/step - accuracy: 0.7808 - loss: 0.7887 - val_accuracy: 0.7811 - val_loss: 0.8017
Epoch 5/30
374/374 ━━━━━━━━━━━━━━━━━━━━ 0s 506us/step - accuracy: 0.8045 - loss: 0.6991 - val_accuracy: 0.7951 - val_loss: 0.7310
Epoch 6/30
374/374 ━━━━━━━━━━━━━━━━━━━━ 0s 510us/step - accuracy: 0.8230 - loss: 0.6337 - val_accuracy: 0.8108 - val_loss: 0.6706
Epoch 7/30
374/374 ━━━━━━━━━━━━━━━━━━━━ 0s 510us/step - accuracy: 0.8359 - loss: 0.5821 - val_accuracy: 0.8252 - val_loss: 0.6288
Epoch 8/30
374/374 ━━━━━━━━━━━━━━━━━━━━ 0s 527us/step - accuracy: 0.8476 - loss: 0.5395 - 

In [22]:
from sklearn.metrics import accuracy_score, recall_score,f1_score,precision_score,confusion_matrix, classification_report
import warnings
warnings.filterwarnings('ignore')

y_pred = np.argmax(model.predict(x_test), axis=1)
test_accuracy = accuracy_score(y_test, y_pred)
prec_best = precision_score(y_test, y_pred, average='macro')
rec_best = recall_score(y_test, y_pred, average='macro')
f1_best = f1_score(y_test, y_pred, average='macro')
print("🔹 Tuned Model Performance:")
print(f"Accuracy: {test_accuracy:.4f}")
print(f"Precision: {prec_best:.4f}")
print(f"Recall: {rec_best:.4f}")
print(f"F1-score: {f1_best:.4f}")
print("\nClassification Report:\n", classification_report(y_test, y_pred))


117/117 ━━━━━━━━━━━━━━━━━━━━ 0s 311us/step
🔹 Tuned Model Performance:
Accuracy: 0.9036
Precision: 0.9061
Recall: 0.9030
F1-score: 0.9036

Classification Report:
               precision    recall  f1-score   support

           0       0.95      0.93      0.94       151
           1       0.81      0.92      0.86       146
           2       0.86      0.95      0.90       142
           3       0.90      0.92      0.91       152
           4       0.87      0.84      0.85       145
           5       0.82      0.88      0.85       150
           6       0.82      0.84      0.83       149
           7       0.87      0.82      0.85       141
           8       0.94      0.86      0.90       105
           9       0.92      0.94      0.93       143
          10       0.88      0.94      0.91       144
          11       0.95      0.89      0.92       134
          12       0.94      0.93      0.93       146
          13       0.93      0.90      0.92       138
          14       0.88    

## 3. Hyperparameter Tuning

In [23]:
from tensorflow import keras
def create_model(activation='relu', optimizer='adam',neurons=64):
    model = keras.Sequential([
        keras.layers.Dense(neurons, input_dim=x_train.shape[1], activation=activation),
         keras.layers.Dense(len(np.unique(y_train)), activation='softmax')
    ])
    model.compile(loss='sparse_categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])
    return model

In [24]:
!pip install scikeras

In [25]:
from scikeras.wrappers import KerasClassifier
keras_clf = KerasClassifier(model=create_model, verbose=0)

In [26]:
param_grid = {
    'model__activation': ['relu', 'tanh'],
    'model__optimizer': ['adam', 'sgd'],
     'model__neurons': [32, 64, 128],
    'batch_size': [16, 32],
    'epochs': [30,50]
}

In [28]:
from sklearn.model_selection import train_test_split,GridSearchCV, RandomizedSearchCV, StratifiedKFold
cv = StratifiedKFold(n_splits=3, shuffle=True, random_state=42)
grid = GridSearchCV(estimator=keras_clf,
                    param_grid=param_grid,
                    scoring='accuracy',
                    cv=cv,   
                    verbose=2)
grid_result = grid.fit(x_train, y_train)

Fitting 3 folds for each of 48 candidates, totalling 144 fits
[CV] END batch_size=16, epochs=30, model__activation=relu, model__neurons=32, model__optimizer=adam; total time=   7.6s
[CV] END batch_size=16, epochs=30, model__activation=relu, model__neurons=32, model__optimizer=adam; total time=   7.4s
[CV] END batch_size=16, epochs=30, model__activation=relu, model__neurons=32, model__optimizer=adam; total time=   7.5s
[CV] END batch_size=16, epochs=30, model__activation=relu, model__neurons=32, model__optimizer=sgd; total time=   6.9s
[CV] END batch_size=16, epochs=30, model__activation=relu, model__neurons=32, model__optimizer=sgd; total time=   6.7s
[CV] END batch_size=16, epochs=30, model__activation=relu, model__neurons=32, model__optimizer=sgd; total time=   6.7s
[CV] END batch_size=16, epochs=30, model__activation=relu, model__neurons=64, model__optimizer=adam; total time=   7.7s
[CV] END batch_size=16, epochs=30, model__activation=relu, model__neurons=64, model__optimizer=adam; 

In [29]:
print("Best parameters:", grid_result.best_params_)
print("Best accuracy:", grid_result.best_score_)

Best parameters: {'batch_size': 16, 'epochs': 50, 'model__activation': 'tanh', 'model__neurons': 128, 'model__optimizer': 'adam'}
Best accuracy: 0.9442212401232087


In [30]:
best_model = grid_result.best_estimator_
y_pred_best = best_model.predict(x_test)
acc_best = accuracy_score(y_test, y_pred_best)
prec_best = precision_score(y_test, y_pred_best, average='macro')
rec_best = recall_score(y_test, y_pred_best, average='macro')
f1_best = f1_score(y_test, y_pred_best, average='macro')
print("🔹 Tuned Model Performance:")
print(f"Accuracy: {acc_best:.4f}")
print(f"Precision: {prec_best:.4f}")
print(f"Recall: {rec_best:.4f}")
print(f"F1-score: {f1_best:.4f}")
print("\nClassification Report:\n", classification_report(y_test, y_pred_best))

🔹 Tuned Model Performance:
Accuracy: 0.9550
Precision: 0.9558
Recall: 0.9546
F1-score: 0.9550

Classification Report:
               precision    recall  f1-score   support

           0       0.97      0.97      0.97       151
           1       0.93      0.93      0.93       146
           2       0.95      0.97      0.96       142
           3       0.96      0.98      0.97       152
           4       0.94      0.94      0.94       145
           5       0.93      0.95      0.94       150
           6       0.93      0.91      0.92       149
           7       0.92      0.93      0.93       141
           8       0.99      0.91      0.95       105
           9       0.92      0.97      0.95       143
          10       0.93      0.94      0.94       144
          11       0.96      0.96      0.96       134
          12       0.99      0.97      0.98       146
          13       0.96      0.96      0.96       138
          14       0.96      0.93      0.95       144
          15    

* Default model accuracy: 0.9036 --> Tuned model accuracy: 0.9550
* Precision, recall, and F1-score also improved after tuning.
* Hyperparameter tuning (neurons, activation, optimizer, learning rate) boosted performance and helped the model generalize better.

In [31]:
# Hyperparameter tuning has a direct positive impact on model performance, making the ANN more accurate and reliable for classifying alphabets.